right/src/macros.c

```diff
@@ -1960,6 +1962,68 @@ static macro_result_t processIfShortcutCommand(bool negate, const char* arg, con
     return processCommand(arg, argEnd);
 }
 
+static macro_result_t processAutoRepeatCommand(const char* arg1, const char* argEnd) {
+    uint32_t wait_until = 0;
+    uint32_t next_delay = AutoRepeatInitialDelay;
+    switch(s->as.autoRepeatPhase) {
+    case AutoRepeatState_Starting:
+        goto run_command;
+    case AutoRepeatState_InitialDelay:
+        wait_until = AutoRepeatInitialDelay;
+        next_delay = AutoRepeatDelayRate;
+        goto process_delay;
+    case AutoRepeatState_Repeating:
+        wait_until = AutoRepeatDelayRate;
+        next_delay = AutoRepeatDelayRate;
+        goto process_delay;
+    }
+
+process_delay:
+    if(s->as.actionActive) {
+        // check for key remains to be pressed
+        bool pendingReleased = PostponerQuery_IsKeyReleased(s->ms.currentMacroKey);
+        bool currentKeyIsActive = currentMacroKeyIsActive();
+        if (!currentKeyIsActive || pendingReleased) {
+            return MacroResult_Finished;
+        }
+        // wait until configured delays
+        if(Timer_GetElapsedTime(&s->as.delayData.start) < wait_until) {
+            return MacroResult_Sleeping;
+        }
+        memset(&s->as.delayData, 0, sizeof s->as.delayData);
+        s->as.actionActive = false;
+    }
+run_command:
+    macro_result_t res = processCommand(arg1, argEnd);
+    if (!(res & MacroResult_ActionFinishedFlag)) {
+        return res;
+    }
+
+    // switch flag
+    switch(s->as.autoRepeatPhase) {
+    case AutoRepeatState_Starting:
+        // check for key hold here as well to avoid fictitiousdelay
```

Comment:
I am actually a bit confused with all the flags and the meaning behind them (I've tried to make the new update be compatible to what I think you meant. Feel free to let me know if otherwise!)

Reply:
I see. Hope following text helps. (I will place it somewhere within the repository for future reference.)

History of the engine:

 - Single-state engine: At first, macro engine executed at most one macro at a time, not even composing with other actions.

   (Macro result: bool indicating either InProgress or ActionFinished)

 - Preemptive scheduler: Single state was quite limiting, so a state pool was introduced and each macro was allowed to execute once during each update cycle. This allowed multiple macros running at a time, and made commands like `holdLayer` or `holdKey` useful. However, free action interleaving caused macros to be prone to race conditions.

   (Macro result: bool indicating either InProgress or ActionFinished)

 - Blocking scheduler: So the macro engine was redesigned so that only one macro state is active at a time, allowing sequences of commands to execute atomicaly. Delay actions and backward jumps would automatically yield, allowing other macros to take turn.

   (Macro result: current flag system described below)

Flags:
 - InProgress, ActionFinished, Done (exactly one of these should be set):

   - InProgress means that an action is not finished and should be called again.

   - ActionFinished means that action has finished. This flag indicates that the outernmost engine call should take care of action epilogue, that is, to load next action and do any other necessary processing.

   - Done is a special flag indicating that action epilogue was already done. This is done by some special commands such as `goTo` which modify the state of the engine.

 - Yield flag indicates that current macro does no longer wish to block others from running. It is returned by delay behaviours (`holdKey, delayUntil, holdLayer,...`) and by backward jumps (because active wait loops).

- Blocking indicates that USB report output needs to be processed.
 
Generally, macro engine returns when either all macros yield, or when a blocking flag is returned, or when execution quota is reached.


Macro sleeping:

- Since macro parsing is stateless, every command has to be reparsed on each invocation. As a consequence, commands like holdLayer can slow down keyboard execution significantly, hindering cursor movement. For this reason, macro sleeping was introduced.

- The idea is that macro can be put to sleep until either some key state changes, or until specified times. 

- The macro is then not executed until the condition is met. Spurious wakeups are allowed, so macro action implementation should always check whether the condition is actually met. (For instance macro may want to sleep until its own key is released. In practice it is also woken up when another key's state changes.)